File: 2016/04.rb
#!/usr/bin/env ruby

class Sector
  attr_reader :id

  def initialize(str)
    caps = str.match(/^(\D+)(\d+)\[(.*)\]/).captures
    @rubish, id, @checksum = caps

    @id = id.to_i
  end

  def to_s
    puts "id       => #{@id} \n" \
         "checksum => #{@checksum} \n" \
         "rubish   => #{@rubish}"
  end

  def valid?
    sorted_rubish = @rubish.gsub(/-/, "")
      .chars
      .sort_by { |c| [-(@rubish.count c), c] }
      .uniq
      .join

    sorted_rubish.start_with? @checksum
  end

  #Caesar shift
  def decrypt
    @rubish.chars.collect { |c|
      if c == "-"
        " "
      else
        tmp = ((c.ord - "a".ord + @id) % 26)
        (tmp + "a".ord).chr
      end
    }.join
  end
end

lines = File.readlines(ARGV.first)
secs = lines.map { |l| Sector.new(l) }.select &:valid?

puts secs.sum &:id
puts secs.find { |s| s.decrypt.include? "north" }.id

File: 2016/04.py
#!/usr/bin/env python3

import re
import numpy


class Room:
    def __init__(self, hash: str, sector: str, checksum: str) -> None:
        self.hash = hash
        self.sector = int(sector)
        self.checksum = checksum

    def isReal(self) -> bool:
        uniq = list(numpy.unique([c for c in re.sub("-", "", self.hash)]))
        uniq.sort(key=lambda x: (-self.hash.count(x), ord(x)))
        checksum = ''.join(uniq)[:5]
        return checksum == self.checksum

    # sort of caesar shift
    def decrypt(self) -> str:
        def shift(c):
            if c == '-':
                return " "
            else:
                tmp = (ord(c) - ord('a') + self.sector) % 26
                return chr(tmp + ord('a'))

        return ''.join([shift(c) for c in self.hash])


rooms = []
for line in open("input/04.txt").readlines():
    matches = re.search("(^[^\d]+)(\d+)\[(\w+)\]", line)
    tmp = Room(matches.group(1), matches.group(2), matches.group(3))
    if tmp.isReal():
        rooms.append(tmp)

print(sum((x.sector for x in rooms)))
print(next(x for x in rooms if re.search("north", x.decrypt())).sector)
